### DISCUSSION TITLE
Deadlock(교착 상태) 현상을 방지하는 방법?

### DISCUSSION THREAD

**Discussion by Hanjaemo** [Hanjaemo/DB-Study#20] [Q&A]
Deadlock(교착 상태) 현상을 방지할 수 있는 방법에는 무엇이 있을까요?

**Comment by gomudayya** [comment]
Deadlock을 방지하는 방법은 트랜잭션 격리수준을 serializable하게 하면 됩니다. 다만 성능적으로 좋지않아 DBMS에서는 데드락 회피기법을 사용한다고 합니다.

데드락 회피 기법중에 우선순위를 이용한 기법을 사용하는데, 자원을 할당할 때 시간 스탬프(Time Stamp)를 사용하여 우선순위를 설정하고, 교착상태가 일어나지 않도록 회피하는 방법으로 Wait-Die 방식과 Wound-Wait 방식이 있습니다. 

여기서 우선순위는 Timestamp값이 작은 것(즉 제일 먼저발생한 트랜잭션이 우선순위입니다)

### Wait-Die

- 먼저 일어난 트랜잭션은 기다리고, 나중에 일어난 트랜잭션은 죽는다. 

  - 여기서 '죽는다' 라는것은 일시정지 시키고, 일정 시간 이후에 다시 실행하는 것 같아요.

  - Tx1이 Tx2가 점유하고 있는 데이터를 요청하면, Tx1은 기다립니다. (앞으로 계속 숫자가 작은것이 먼저 일어났다고 가정할게요)

  - Tx3이 Tx2가 점유하고 있는 데이터를 요청하면 Tx3는 죽습니다. 

- Tx1과 Tx2가 데드락이 발생하여 서로의 리소스가 락에서 풀리기를 기다리고 있다고 해봅시다.

- Tx1은 먼저 일어난 트랜잭션이므로 기다리고, Tx2는 나중에 일어난 트랜잭션이므로 죽습니다. Tx2에 의해 걸려있던 락이 해제되고 Tx1이 실행됩니다. 이후에 Tx2가 다시 복귀하여 실행됩니다. 
 
### Wound-Wait

- 먼저 일어난 트랜잭션은 빼앗고(선점), 나중에 일어난 트랜잭션은 기다린다.

  - Tx1이 Tx2가 점유하고 있는 데이터를 요청하면 Tx2는 일시정지 되고, Tx1이 데이터를 선점합니다.

- 마찬가지로 Tx1과 Tx2가 데드락이 발생하여 서로의 리소스가 락에서 풀리기를 기다리고 있다고 해봅시다.

- Tx1이 먼저 일어난 트랜잭션이므로 선점해서 먼저 실행되고, 이후에 Tx2가 실행됩니다.
 

<br>
DBMS 내부적으로 이미 구현되어 있는 내용이라 개발하면서 크게 신경 쓸 내용은 아닌 것 같긴 합니다.
<br>

참고 링크
- https://www.youtube.com/watch?v=5yobcvhMV2M
- https://www.youtube.com/watch?v=vGytqg8KqCY
- https://blog.naver.com/ndb796/221243161017
- https://parkmuhyeun.github.io/etc/database/2022-07-03-Deadlock/
